### PR TITLE
broadcast[!] over combinations of scalars and sparse vectors/matrices

### PR DESCRIPTION
**Pull request by Sacha0**
(Requires #19667, #19723, and #19690, which form the first four commits / most of the diff.)

This pull request extends sparse `broadcast[!]` to handle combinations of scalars and sparse vectors/matrices. **The tl;dr good**: It works, and with #19667/#19273 the overall return el/type is inferred consistently. **The tl;dr not so good**: Inference seems unhappy in another way in some cases, impacting performance. The details:

This PR's general approach is to: (1) capture the scalar arguments to `broadcast[!]` in a closure and simultaneously collect the sparse vector/matrix arguments; and then (2) pass that closure and the sparse vector/matrix arguments to existing `broadcast[!]` methods that accept solely sparse vector/matrix arguments.

With #19667/#19723, `_return_type`/`_broadcast_eltype` is inferable and yields the correct result eltype when called on the closure and collected sparse vector/matrix arguments described above. Hence the overall return el/type is correct. Unfortunately, in some cases inference seems unable to determine the return type of direct calls to that closure, in those cases causing allocation in both the `broadcast[!]` entry points and the underlyin routines (`_map_zeropres!`, `_broadcast_zeropres!`) and lackluster performance. (Will try to construct an MRE at some point.)

This solution's general finickiness drives me to the conclusion that we should instead handle scalar arguments directly in the underlying routines. I do not know whether I will have time for another rewrite of the underlying routines before feature freeze. So I suggest we incorporate this (correct but performance-suboptimal) implementation before feature freeze, allowing another underlying routine rewrite for performance after feature freeze. Best!

(Edit: Also the closure construction allocates a bit, not sure why. Pointers much appreciated.)

### CODE REVIEW

**Comment by nalimilan** [comment]
Do inference failures affect common cases or only very specific ones? As long as the behavior is correct, we could merge this before the feature freeze, and improve performance later.

File: base/sparse/higherorderfns.jl [hunk 1]
```diff
@@ -0,0 +1,889 @@
...
+@inline capturescalars(f, passedargstup, nonscalararg::SparseVecOrMat) =
+    (passlastnonscalar(f), (passedargstup..., nonscalararg))
+@inline passlastnonscalar(f) = (passed...) -> f(Base.front(passed), (last(passed),))
+@inline capturelastscalar(f, scalararg) = (passed...) -> f(passed, (scalararg,))
+
+# NOTE: The following two method definitions work around #19096.
+broadcast{Tf,T}(f::Tf, ::Type{T}, A::SparseMatrixCSC) = broadcast(y -> f(T, y), A)
+broadcast{Tf,T}(f::Tf, A::SparseMatrixCSC, ::Type{T}) = broadcast(x -> f(x, T), A)
+
+end
```
Comment (stevengj) [inline base/sparse/higherorderfns.jl]: missing newline

File: base/sparse/higherorderfns.jl [hunk 2]
```diff
@@ -0,0 +1,889 @@
...
+    _initrowforcol_all(j, rowsentinel, tail(isemptys), tail(expandsverts), tail(ks), tail(As))...)
+@inline _defargforcol(j, isempty, expandsvert, k, A) =
+    expandsvert && !isempty ? storedvals(A)[k] : zero(eltype(A))
+@inline _defargforcol_all(j, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::Tuple{}) = ()
+@inline _defargforcol_all(j, isemptys, expandsverts, ks, As) = (
+    _defargforcol(j, first(isemptys), first(expandsverts), first(ks), first(As)),
+    _defargforcol_all(j, tail(isemptys), tail(expandsverts), tail(ks), tail(As))...)
+# fusing the following defs. avoids a few branches and construction of a tuple, yielding 1-20% runtime reduction
+# @inline _isactiverow(activerow, row) = row == activerow
+# @inline _isactiverow_all(activerow, ::Tuple{}) = ()
```
Comment (stevengj) [inline base/sparse/higherorderfns.jl]: What are all of these commented-out definitions?

File: base/sparse/higherorderfns.jl [hunk 3]
```diff
@@ -0,0 +1,889 @@
...
+# @inline _updateind_all(activerows, ks) = (
+#     _updateind(first(activerows), first(ks)),
+#     _updateind_all(tail(activerows), tail(ks))...)
+# @inline _updaterow(rowsentinel, isrowactive, presrow, k, stopk, A) =
+#     isrowactive ? (k < stopk ? storedinds(A)[k] : oftype(presrow, rowsentinel)) : presrow
+# @inline _updaterow_all(rowsentinel, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::Tuple{}) = ()
+# @inline _updaterow_all(rowsentinel, activerows, rows, ks, stopks, As) = (
+#     _updaterow(rowsentinel, first(activerows), first(rows), first(ks), first(stopks), first(As)),
+#     _updaterow_all(rowsentinel, tail(activerows), tail(rows), tail(ks), tail(stopks), tail(As))...)
+@inline function _fusedupdatebc(rowsentinel, activerow, row, defarg, k, stopk, A)
```
Comment (stevengj) [inline base/sparse/higherorderfns.jl]: Some comment on what `_fusedupdate` is supposed to do would be helpful.

File: base/sparse/higherorderfns.jl [hunk 4]
```diff
@@ -0,0 +1,889 @@
...
+@inline function broadcast!_c{N}(f, ::Type{AbstractSparseArray}, dest::SparseVecOrMat, mixedsrcargs::Vararg{Any,N})
+    parevalf, passedsrcargstup = capturescalars(f, mixedsrcargs)
+    return broadcast!(parevalf, dest, passedsrcargstup...)
+end
+# capturescalars takes a function (f) and a tuple of mixed sparse vectors/matrices and
+# broadcast scalar arguments (mixedargs), and returns a function (parevalf) and a reduced
+# argument tuple (passedargstup) containing only the sparse vectors/matrices in mixedargs
+# in their orginal order, and such that the result of broadcast(g, passedargstup...) is
+# broadcast(f, mixedargs...)
+@inline capturescalars(f, mixedargs) =
```
Comment (stevengj) [inline base/sparse/higherorderfns.jl]: Is this useful to do for non-sparse `broadcast` as well?

**Comment by stevengj** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

File: base/sparse/higherorderfns.jl (hunk 1 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Fixed in #19690. (Commits three and four are #19690.) Thanks!

File: base/sparse/higherorderfns.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: The commented-out definitions are what I fused into `_fusedupdatebc` (for performance) per the comment on line 797. I left the unfused definitions in place (along with commented out calls to those methods before calls to `_fusedupdatebc`) as an easier-to-follow form of the fused version. Best!

File: base/sparse/higherorderfns.jl (hunk 3 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Please see my response above. Best!

File: base/sparse/higherorderfns.jl (hunk 4 above)
Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Were the inference and allocation issues impacting performance (see [the original post](https://github.com/JuliaLang/julia/pull/19724#issue-197645527) and [this comment in test/sparse/higherorderfns.jl:244](https://github.com/JuliaLang/julia/pull/19724/files/210512275e98b81b255d5b9ca124513b3cbcda81..20fe9fa35804c5f758cf8355380ac1c6798841fd#diff-cc2da18954fb25a7e3d7ad8f80b6fed9R224)) fixed, possibly. Though as in the original post's last paragraph, I plan to head the other direction for now. Best!

**Comment by Sacha0** [comment]
> Do inference failures affect common cases or only very specific ones?

Most cases involving one or two sparse vector/matrix arguments and a small number of scalars seem to avoid the inference issue, though they do hit the (much less significant) issue (1) mentioned in [this comment in test/sparse/higherorderfns.jl:244](https://github.com/JuliaLang/julia/pull/19724/files/210512275e98b81b255d5b9ca124513b3cbcda81..20fe9fa35804c5f758cf8355380ac1c6798841fd#diff-cc2da18954fb25a7e3d7ad8f80b6fed9R224)). Most cases involving three or more sparse vector/matrix arguments hit the inference issue.

> As long as the behavior is correct, we could merge this before the feature freeze, and improve performance later.

:)

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19724#issuecomment-269388791) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/94cb5b3de893699e9dc75d375ebb78a23f35b5fd/d7e8303_vs_915a07d/report.md). cc @jrevels

**Comment by stevengj** [comment]
The nanosoldier regression seems to be the usual `floatexp` test noise, so it should be fine to merge.

**Comment by tkelman** [comment]
the pr's that make up the first several commits should be finished first, otherwise we're splitting review of them in multiple places

**Comment by Sacha0** [comment]
Rebased out commits from merged pull requests. Now depends only on #19723 or #19787+#19723. Best!

File: test/sparse/higherorderfns.jl [hunk 5]
```diff
@@ -226,8 +226,29 @@ end
...
+            @test_broken (@allocated broadcast!(*, X, sparseargs...)) == 0
+            # This test (and the analog below) fails for three reasons:
+            # (1) In all cases, generating the closures that capture the scalar arguments
+            #   results in allocation, not sure why.
+            # (2) In some cases, though _broadcast_eltype (which wraps _return_type)
+            #   consistently provides the correct result eltype when passed the closure
+            #   that incorporates the scalar arguments to broadcast (and, with #19667,
+            #   is inferable, so the overall return type from broadcast is inferred),
+            #   in some cases inference seems unable to determine the return type of
+            #   direct calls to that closure. This issue causes variables in in both the
```
Comment (tkelman) [inline test/sparse/higherorderfns.jl]: in in duplicated

File: test/sparse/higherorderfns.jl [hunk 6]
```diff
@@ -226,8 +226,29 @@ end
...
+            #   broadcast[!] entry points (fofzeros = f(_zeros_eltypes(args...)...)) and
+            #   the driver routines (Cx in _map_zeropres! and _broadcast_zeroprs!) to have
```
Comment (tkelman) [inline test/sparse/higherorderfns.jl]: zeroprs missing a second e?

File: test/sparse/higherorderfns.jl (hunk 5 above)
Comment (Sacha0) [inline test/sparse/higherorderfns.jl]: Fixed on push. Thanks!

File: test/sparse/higherorderfns.jl (hunk 6 above)
Comment (Sacha0) [inline test/sparse/higherorderfns.jl]: Fixed on push. Thanks!

**Comment by Sacha0** [comment]
Rebased and replaced the first commit with the latest version of #19723. Best!

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19724#issuecomment-269900212) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/220bcff1f8c87753c9f2669ea94a36b008c59065/cef1c0f_vs_9cc7815/report.md). cc @jrevels

File: base/sparse/higherorderfns.jl [hunk 7]
```diff
@@ -837,15 +839,52 @@ end
...
+    return broadcast(parevalf, passedargstup...)
+end
+@inline function broadcast!_c{N}(f, ::Type{AbstractSparseArray}, dest::SparseVecOrMat, mixedsrcargs::Vararg{Any,N})
+    parevalf, passedsrcargstup = capturescalars(f, mixedsrcargs)
+    return broadcast!(parevalf, dest, passedsrcargstup...)
+end
+# capturescalars takes a function (f) and a tuple of mixed sparse vectors/matrices and
+# broadcast scalar arguments (mixedargs), and returns a function (parevalf) and a reduced
+# argument tuple (passedargstup) containing only the sparse vectors/matrices in mixedargs
+# in their orginal order, and such that the result of broadcast(g, passedargstup...) is
```
Comment (tkelman) [inline base/sparse/higherorderfns.jl]: is `g` the same as `parevalf` ?

Comment (Sacha0) [inline base/sparse/higherorderfns.jl]: Yes, and par-eval as in partial evaluation. The former fixed and the latter clarified on push. Thanks!

**Comment by Sacha0** [comment]
Absent objections or requests for time, I plan to merge this Monday morning PST. Best!

File: base/broadcast.jl [hunk 8]
```diff
@@ -202,7 +202,9 @@ Note that `dest` is only used to store the result, and does not supply
 arguments to `f` unless it is also listed in the `As`,
 as in `broadcast!(f, A, A, B)` to perform `A[:] = broadcast(f, A, B)`.
 """
-@inline function broadcast!{N}(f, C::AbstractArray, A, Bs::Vararg{Any,N})
+@inline broadcast!{N}(f, C::AbstractArray, A, Bs::Vararg{Any,N}) =
+    broadcast!_c(f, containertype(C, A, Bs...), C, A, Bs...)
+@inline function broadcast!_c{N}(f, ::Type, C::AbstractArray, A, Bs::Vararg{Any,N})
```
Comment (pabloferz) [inline base/broadcast.jl]: Would `broadcast_c!` be better?

Comment (Sacha0) [inline base/broadcast.jl]: No preference; I can see arguments for each form. Thoughts? Thanks!

**Comment by pabloferz** [comment]
> This solution's general finickiness drives me to the conclusion that we should instead handle scalar arguments directly in the underlying routines.

I'm somewhat hesitating between following that approach or implementing some kind of optimization like the one in this PR for broadcasting methods in general. For one, this PR's mechanism is a way to avoid problems with `::Type` arguments which are somewhat problematic (see #19849 and #19879). On the other hand, the `Broadcast` approach (handling scalars directly in the routines) would probably avoid some of the problems you saw with inference. I'm not sure (yet) if a mixed approach between the two solutions is plausible.

For now I'd go with this solution anyway and would eventually revisit other options.

File: base/broadcast.jl (hunk 8 above)
Comment (pabloferz) [inline base/broadcast.jl]: I like `broadcast_c!` better, but however you prefer is fine.

Comment (Sacha0) [inline base/broadcast.jl]: Changed `broadcast!_c` to `broadcast_c!` throughout. Thanks!

**Comment by Sacha0** [comment]
Thanks all!

**Comment by Sacha0** [comment]
Master is failing tests as of this pull request (particularly those introduced by this pull request). I conjecture that #19922 (merged after this pull request's last CI run) broke this pull request. Should have a fix shortly if the former is true. Best!

**Pull request merged**